src/app/fdctl/run/tiles/fd_repair.c

```diff
@@ -0,0 +1,656 @@
+             ushort                 dst_port,
+             uchar const *          payload,
+             ulong                  payload_sz,
+             ulong                  tsorig ) {
+  uchar * packet = fd_chunk_to_laddr( ctx->net_out_mem, ctx->net_out_chunk );
+
+  eth_ip_udp_t * hdr = (eth_ip_udp_t *)packet;
+  uchar mac[6] = {0};
+  populate_packet_header_template( hdr, payload_sz, ctx->repair_my_intake_addr.addr, mac, ctx->repair_intake_listen_port );
+
+  hdr->udp->net_dport = dst_port;
+
+  memcpy( hdr->eth->dst, mac, 6UL );
+  memcpy( hdr->ip4->daddr_c, &dst_ip_addr, 4UL );
+
+  hdr->ip4->check = fd_ip4_hdr_check( ( fd_ip4_hdr_t const *)FD_ADDRESS_OF_PACKED_MEMBER( hdr->ip4 ) );
+
+  ulong packet_sz = payload_sz + sizeof(eth_ip_udp_t);
+  fd_memcpy( packet+sizeof(eth_ip_udp_t), payload, payload_sz );
+
+  hdr->udp->check = fd_ip4_udp_check( *(uint *)FD_ADDRESS_OF_PACKED_MEMBER( hdr->ip4->saddr_c ), 
+                                      *(uint *)FD_ADDRESS_OF_PACKED_MEMBER( hdr->ip4->daddr_c ), 
+                                      (fd_udp_hdr_t const *)FD_ADDRESS_OF_PACKED_MEMBER( hdr->udp ), 
+                                      packet + sizeof(eth_ip_udp_t) );
+
+  ulong tspub = fd_frag_meta_ts_comp( fd_tickcount() );
+  ulong sig = fd_disco_netmux_sig( ctx->repair_my_intake_addr.addr, ctx->repair_intake_listen_port, dst_ip_addr, DST_PROTO_OUTGOING, FD_NETMUX_SIG_MIN_HDR_SZ );
+  fd_mcache_publish( ctx->net_out_mcache, ctx->net_out_depth, ctx->net_out_seq, sig, ctx->net_out_chunk, packet_sz, 0UL, tsorig, tspub );
+  ctx->net_out_seq   = fd_seq_inc( ctx->net_out_seq, 1UL );
+  ctx->net_out_chunk = fd_dcache_compact_next( ctx->net_out_chunk, packet_sz, ctx->net_out_chunk0, ctx->net_out_wmark );
+}
+
+static inline void
+handle_new_cluster_contact_info( fd_repair_tile_ctx_t * ctx,
+                                 uchar const *          buf ) {
+  ulong const * header = (ulong const *)fd_type_pun_const( buf );
+
+  ulong dest_cnt = FD_LOAD( ulong, header );
+
+  if( dest_cnt >= MAX_REPAIR_PEERS ) {
+    FD_LOG_WARNING(( "Cluster nodes had %lu destinations, which was more than the max of %lu", dest_cnt, MAX_REPAIR_PEERS ));
+    return;
+  }
+
+  fd_shred_dest_wire_t const * in_dests = (fd_shred_dest_wire_t const *)( header+1UL );
+
+  for( ulong i=0UL; i<dest_cnt; i++ ) {
+    fd_repair_peer_addr_t repair_peer = {
+      .addr = in_dests[i].ip4_addr,
+      .port = in_dests[i].udp_port,
+    };
+   
+    fd_repair_add_active_peer( ctx->repair, &repair_peer, in_dests[i].pubkey );
+  }
+}
+
+static inline void
+handle_new_repair_requests( fd_repair_tile_ctx_t * ctx,
+                            uchar const *          buf,
+                            ulong                  buf_sz ) {
+
+  fd_repair_request_t const * repair_reqs = (fd_repair_request_t const *) buf;
+  ulong repair_req_cnt = buf_sz;
+
+  for( ulong i=0; i<repair_req_cnt; i++ ) {
+    fd_repair_request_t const * repair_req = &repair_reqs[i];
```

Comment:
You could read out of bounds here if the buf_sz < this offset. Can you validate the reads

Reply:
I don't understand what's going on here. Are you saying the store tile is sending frags with the meaning of the `sz` field changed to be the number of requests rather than the actual fragment size? Even if you did that (which you shouldn't), it's still wrong - you aren't validating that the memory you are reading here would be in bounds. I can trivially construct a fragment (eg, by sending you a packet with sz field of ULONG_MAX) which crashes your tile.